x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleType.java

```diff
@@ -226,5 +255,24 @@ public void validate(Collection<Phase> phases) {
                 "] action may not be used in the [" + HOT_PHASE +
                 "] phase without an accompanying [" + RolloverAction.NAME + "] action");
         }
+
+        // look for phases that have the migrate action enabled and also specify allocation rules via the AllocateAction
+        String phasesWithConflictingMigrationActions = phases.stream()
+            .filter(phase -> phase.getActions().containsKey(MigrateAction.NAME) &&
+                ((MigrateAction) phase.getActions().get(MigrateAction.NAME)).isEnabled() &&
+                phase.getActions().containsKey(AllocateAction.NAME) &&
+                definesAllocationRules((AllocateAction) phase.getActions().get(AllocateAction.NAME))
+            )
+            .map(Phase::getName)
+            .collect(Collectors.joining(","));
+        if (Strings.hasText(phasesWithConflictingMigrationActions)) {
+            throw new IllegalArgumentException("phases [" + phasesWithConflictingMigrationActions + "] specify an enabled " +
+                MigrateAction.NAME + " action and the " + AllocateAction.NAME + " action. specify only one data migration action in these" +
+                " phases");
+        }
+    }
+
+    private boolean definesAllocationRules(AllocateAction action) {
+        return action.getRequire() != null || action.getInclude() != null || action.getExclude() != null;
```

Comment:
These should use the `.isEmpty()` check rather than checking for `null`, since they could be `"include": {}` and be non-null but not actually have any allocation rules